### DISCUSSION TITLE
Threshold by width and height?

### DISCUSSION THREAD

**Discussion by SB2020-eye** [ImageMagick/ImageMagick#2087] [Help]
Hello.

I'm new to ImageMagick and would greatly appreciate anyone's input.

In a binary image, is there a way to get rid of objects that don't fall within a certain width and certain height? (Ie, set a min and max width and a min and max height)

I've experimented with connected component and area threshold. These are really great. Still, I am needing to exclude objects based on specific widths and heights, as opposed to area per se.

Thanks for your consideration. :)

**Comment by dlemstra** [comment]
Could you please include some example images to make it more clear what you want to accomplish?

**Comment by fmw42** [comment]
That was a thought I had as well, but it never got implemented. But you can extract the width and height of the object from its bounding box (WxH+X+Y) and then filter on those subsequently. Or you can use the approximations from connected-components:major-axis-threshold and connected-components:minor-axis-threshold if your regions are horizontal or vertically oriented. But it would be best if you post and example image and provide your IM version and platform. If on Unix-like system, I can show you how to filter on W and H from the bounding box. Sorry, but I do not use Windows and scripting is different.

**Comment by SB2020-eye** [comment]
Thank you both for responding. I'll lay out what I'm seeking to do, including some examples.

I'm on Windows and using ImageMagick 7.0.10-16 Q16 x64 2020-05-31.

From images like this:
![sample](https://user-images.githubusercontent.com/56878280/83545666-84b61d80-a4cd-11ea-9200-2d90a1c8219f.png)

I wish to get image files of _all_ letters - like this:
![a1](https://user-images.githubusercontent.com/56878280/83546990-8c76c180-a4cf-11ea-92d2-929aef93e258.png)
![a2](https://user-images.githubusercontent.com/56878280/83546991-8c76c180-a4cf-11ea-933c-c5d8ab265b87.png)
![b1](https://user-images.githubusercontent.com/56878280/83546988-8c76c180-a4cf-11ea-900f-8bbb70ebe552.png)
These need to retain color and resolution. I need _all_ as, bs, etc (not just one example of each). I don't need letters from the annotations (the smaller script between lines). (I _would_  eventually like to have the larger, ornate letters - but I anticipate that will require a separate process.)

This is relatively new for me. It could be that connected components isn't the best way to go. (And please feel free to let me know if that's the case! :) ) But I'll show you why I asked my original question. When I crop the borders of the image at top, binarize, invert, and run

```
magick ^
  sample.png ^
  -define connected-components:verbose=true ^
  -define connected-components:area-threshold=900-30000 ^
  -define connected-components:mean-color=true ^
  -connected-components 4 ^
  -threshold 50%% ^
  sampleCC.png
```
I get
![page101vCC](https://user-images.githubusercontent.com/56878280/83548120-3c98fa00-a4d1-11ea-9a05-79d19e71f64f.png)
Being able to be specific about width and height parameters (as opposed to area per se) would help remove the remaining, unwanted annotations you can still see. (Well, thinking about it, since so many letters are connected, perhaps only height makes a difference in this case.)

I also am thinking that width and height could help in separating individual letters - which I'm still not sure how I'll go about exactly. For instance, I have tried out the code from a project I discovered online that utilizes stroke width (eg, the width of an "i") and letter width and height (mins and maxes) in order to isolate and clip out each letter. It does a very good job. However, the output lowers the resolution, leaves a colored mask over the letter, and only processes up to a certain amount of letters and then stops. If I could code, perhaps I could just alter it (which the author has given me permission to do). But I've tried and I'm clueless. Anyways, I do understand that he gets his individual letters clipped by taking advantage of the handwriting's conformity of stroke width and letter dimensions. (If interested, his paper describing some of this is here: [http://ceur-ws.org/Vol-2314/paper2.pdf](url).)

All this said, I am open to any means towards the end! :)

(PS I also note that connected components fills holes like the hole in letter "e"s, which isn't wanted.)
